brave/utility/tor/tor_impl.cc

```diff
@@ -0,0 +1,171 @@
+
+#if defined(OS_POSIX)
+int pipehack[2];
+
+static void SIGCHLDHandler(int signal) {
+  pid_t p;
+  int status;
+
+  if ((p = waitpid(-1, &status, WNOHANG | WUNTRACED)) != -1) {
+    if (WIFSIGNALED(status) || WCOREDUMP(status)) {
+      write(pipehack[1], &p, sizeof(pid_t));
+    } else if (WIFEXITED(status)) {
+      // TODO(darkdh): Handle "Address already in use" issue. Not notifying as
+      // workaroud for now to prevent infinitely relaunch
+      if (!WEXITSTATUS(status))
+        write(pipehack[1], &p, sizeof(pid_t));
+    }
+  }
+}
+
+static void SetupPipeHack() {
+  if (pipe(pipehack) == -1)
+    LOG(ERROR) << "pipehack";
+
+  int flags;
+  for (size_t i = 0; i < 2; ++i) {
+    if ((flags = fcntl(pipehack[i], F_GETFL)) == -1)
+      LOG(ERROR) << "get flags";
+    if (i == 1)
+      flags |= O_NONBLOCK;
+    flags |= O_CLOEXEC;
+    if (fcntl(pipehack[i], F_SETFL, flags) == -1)
+      LOG(ERROR) << "set flags";
+  }
+
+  struct sigaction action;
+  memset(&action, 0, sizeof(action));
+  action.sa_handler = SIG_IGN;
+  sigaction(SIGPIPE, &action, NULL);
+
+  memset(&action, 0, sizeof(action));
+  action.sa_handler = SIGCHLDHandler;
+  sigaction(SIGCHLD, &action, NULL);
+}
+#endif
+
+namespace brave {
+
+class TorLauncherDelegate : public base::LaunchOptions::PreExecDelegate {
+ public:
+    TorLauncherDelegate() {}
+    ~TorLauncherDelegate() override {}
+
+    void RunAsyncSafe() override {
+#if defined(OS_POSIX)
+      for (size_t i = 0; i < 2; ++i)
+        close(pipehack[i]);
+#endif
+      // TODO(darkdh): doesn't prevent ctrl+c propagated by terminal
+      struct sigaction action;
+      memset(&action, 0, sizeof(action));
+      action.sa_handler = SIG_IGN;
+      sigaction(SIGINT, &action, NULL);
+    }
+ private:
+    DISALLOW_COPY_AND_ASSIGN(TorLauncherDelegate);
+};
+
+TorLauncherImpl::TorLauncherImpl(
+    std::unique_ptr<service_manager::ServiceContextRef> service_ref)
+    : service_ref_(std::move(service_ref)) {
+  SetupPipeHack();
+}
+
+TorLauncherImpl::~TorLauncherImpl() {
+  if (tor_process_.IsValid()) {
+#if defined(OS_POSIX)
+    for (size_t i = 0; i < 2; ++i)
+      close(pipehack[i]);
+#endif
+    tor_process_.Terminate(0, false);
+#if defined(OS_MACOSX)
+    base::PostTaskWithTraits(
+        FROM_HERE, {base::MayBlock(), base::TaskPriority::BACKGROUND},
+        base::BindOnce(&base::EnsureProcessTerminated, Passed(&tor_process_)));
+#else
+    base::EnsureProcessTerminated(std::move(tor_process_));
+#endif
+  }
+}
+
+void TorLauncherImpl::Launch(const base::FilePath& tor_bin,
+                             const std::vector<std::string>& args,
+                             LaunchCallback callback) {
+  base::CommandLine cmdline(tor_bin);
+  for (size_t i = 0; i < args.size(); ++i) {
+    cmdline.AppendArg(args[i]);
+  }
+  base::LaunchOptions launchopts;
+  TorLauncherDelegate tor_launcher_delegate;
+  launchopts.pre_exec_delegate = &tor_launcher_delegate;
```

Comment:
What's the motivation for using this pre-exec delegate logic instead of just installing the signal handler right here?

Reply:
this can be removed. it was using to install SIGINT handler on child.

Reply:
I see -- does the pre-exec delegate run in the child, after fork but before exec?  We should make sure that all file descriptors not relevant to tor's needs are closed in the child, and maybe that's the place you want to call setsid.

Reply:
yeah after fork but before exec exactly!

Reply:
fixed in fea7dca